plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt

```diff
@@ -231,47 +236,82 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName, val isInterface:
+    private fun initSharedPredicate(): Predicate {
+        val filteredFields = fields.values
+            .filterIsInstance<UserFieldEmbedding>()
+            .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+
+        val accessFields = filteredFields.map { FieldAccessTypeInvariantEmbedding(it, PermExp.WildcardPerm()).fillHole(subjectEmbedding) }
 
-        // For the moment ALWAYS_WRITEABLE fields with some class type do not exist.
-        // Whether to include them here will need to be considered in the future.
         // We need to pass in the field access since the predicates are simply the ones for the type.
-        val accessFieldPredicates = fields.values
-            .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
-            .mapNotNull { it.type.predicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it)) }
+        val accessFieldPredicates = filteredFields
+            .mapNotNull { it.type.sharedPredicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it)) }
 
-        val fieldsProvenInvariants = fields.values
-            .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+        val fieldsProvenInvariants = filteredFields
             .flatMap { it.type.provenInvariants().fillHoles(PrimitiveFieldAccess(subjectEmbedding, it)) }
 
-        val accessSuperTypesPredicates = superTypes
-            .filterIsInstance<ClassTypeEmbedding>()
+        val accessSuperTypesPredicates = classSuperTypes
             .map { PredicateAccessTypeInvariantEmbedding(it.name, PermExp.WildcardPerm()).fillHole(subjectEmbedding) }
 
         val body = (accessFields + accessFieldPredicates + accessSuperTypesPredicates + fieldsProvenInvariants).toConjunction()
         return Predicate(name, listOf(subjectEmbedding.toLocalVarDecl()), body.pureToViper(toBuiltin = true))
     }
 
+    private fun initUniquePredicate(): Predicate {
+        val filteredFields = fields.values.filterIsInstance<UserFieldEmbedding>()
+
+        val accessFields = filteredFields.map {
+            if (it.symbol.isVal) {
+                FieldAccessTypeInvariantEmbedding(it, PermExp.WildcardPerm()).fillHole(subjectEmbedding)
+            } else {
+                FieldAccessTypeInvariantEmbedding(it, PermExp.FullPerm()).fillHole(subjectEmbedding)
+            }
+        }
+
+        val accessFieldSharedPredicates = filteredFields
+            .mapNotNull { it.type.sharedPredicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it)) }
+
+        val accessFieldUniquePredicates = filteredFields
+            .filter { it.isUnique }
+            .mapNotNull { it.type.uniquePredicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it)) }
+
+        val fieldsProvenInvariants = filteredFields
+            .flatMap { it.type.provenInvariants().fillHoles(PrimitiveFieldAccess(subjectEmbedding, it)) }
```

Comment:
Is there a case that this isn't already present in the shared predicate?

Reply:
`var` fields are not included in the shared predicates.
`provenInvariants` of `val` fields appear both in the shared and the unique predicate, but it shouldn't be a problem.
Do you think it can be worth to remove `provenInvariants` of `val` fields from the unique predicate?